src/kanata.rs

```diff
@@ -253,20 +310,78 @@ impl Kanata {
         Ok(())
     }
 
-    fn check_handle_layer_change(&mut self) {
+    fn check_handle_layer_change(&mut self, tx: &Sender<EventNotification>) {
         let cur_layer = self.layout.current_layer();
         if cur_layer != self.prev_layer {
+            let new = self.layer_info[cur_layer].name.clone();
             self.prev_layer = cur_layer;
             self.print_layer(cur_layer);
+
+            match tx.try_send(EventNotification::LayerChange { new }) {
+                Ok(_) => {}
+                Err(error) => {
+                    log::error!("could not sent event notification: {}", error);
+                }
+            }
         }
     }
 
     fn print_layer(&self, layer: usize) {
-        log::info!("Entered layer:\n{}", self.layer_strings[layer]);
+        log::info!("Entered layer:\n{}", self.layer_info[layer].cfg_text);
+    }
+
+    pub fn start_notification_loop(kanata: Arc<Mutex<Self>>, rx: Receiver<EventNotification>) {
+        info!("Kanata: entering the event notification loop");
+        std::thread::spawn(move || {
+            loop {
+                match rx.recv() {
+                    Err(_) => {
+                        panic!("channel disconnected")
+                    }
+                    Ok(event) => {
+                        let k = kanata.lock();
```

Comment:
I don't like that we're acquiring a lock on the Kanata struct here; it means the processing loop is still dependent on and could be blocked by TCP operations.

It seems to me that the NotificationServer doesn't need to be part of the Kanata struct, so it would be better to move it out to avoid having lock contention between TCP operations and the keyboard operations.